extensions/parquet/table/src/main/java/io/deephaven/parquet/table/pagestore/OffsetIndexBasedColumnChunkPageStore.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.table.pagestore;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.base.verify.Require;
+import io.deephaven.chunk.attributes.Any;
+import io.deephaven.engine.page.ChunkPage;
+import io.deephaven.parquet.table.pagestore.topage.ToPage;
+import io.deephaven.parquet.base.ColumnChunkReader;
+import io.deephaven.parquet.base.ColumnPageReader;
+import org.apache.parquet.internal.column.columnindex.OffsetIndex;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+
+final class OffsetIndexBasedColumnChunkPageStore<ATTR extends Any> extends ColumnChunkPageStore<ATTR> {
+    private final OffsetIndex offsetIndex;
+    private final int numPages;
+    /**
+     * Set if first ({@link #numPages}-1) pages have equal number of rows
+     */
+    private boolean isPageSizeFixed;
+    /**
+     * Fixed number of rows per page, only valid if {@link #isPageSizeFixed} is true. Used to map from row index to page
+     * number.
+     */
+    private final long fixedPageSize;
+    private final Object[] objectsForSynchronizingPageAccess;
+    private final ColumnPageReader[] columnPageReaders;
+    private final WeakReference<PageCache.IntrusivePage<ATTR>>[] pages;
```

Comment:
I wonder if it's better to organize this w/ a single array, potentially do away w/ the Object.

```java
class PageState {
  ColumnPageReader reader;
  WeakReference<PageCache.IntrusivePage<ATTR>> page;
}
private final PageState[] states;
```

I think there's a two-stage sync model where you wouldn't even have to allocate the states up front:

```java
PageState get(int i) {
  if (states[i] == null) {
    // global sync if not allocated
    synchronize(this) {
      if (states[i] == null) {
        states[i] = new PageState();
      }
    }
  }
}

class PageState {
  ...
  PageCache.IntrusivePage<ATTR> get() {
    PageCache.IntrusivePage<ATTR> localPage;
    if ((localPage = page) == null)) {
      synchronize(this) {
          if ((localPage = page) == null)) {
           ...
         }
      }
      return localPage;
    }
```

I think this is better than maintaining Object sidecar model.

Reply:
Does `ColumnPageReader` need to be a hard ref? In this model I posted, can we have `WeakRef<PageState>` instead?

Reply:
No, ColumnPageReader doesn't need to be a hard ref. In fact, we are thinking of completely removing the stored reference to the reader object. This comment has the details: https://github.com/deephaven/deephaven-core/pull/4844#discussion_r1396299390

So in the latest patch, I have removed storing reference to the reader.